[ADD] shopfloor_packing_info: Messages for Checkout Packing Info

carlosserratoro opened the pull request:
This PR adds the option to choose predefined messages for the
parameter "Checkout Packing Info", that is set on the res.partner
and reused in other places. For this, the original text field
shopfloor_packing_info has been transformed into a Many2one that
points to the new model shopfloor.packing.info. Thus, the field
in the res.partner has been renamed to shopfloor_packing_info_id,
that has caused some renames in other files that used it. These
new messages are archived instead of deleted, to keep track of
old messages associated to existing (or archived) res.partners.

This new model, shopfloor.packing.info has been moved to the new
module shopfloor_packing_info. Migration hooks have been provided
to migrate the data, that will happen if the old module is updated
at the same time in which the new module is installed.

The new predefined messages can be created directly in the res.partner
or in the menu Inventory > Settings > Shopfloor.

To migrate the data, the existing data has been migrated by assuming
that the first line is the name while all the content is for the text
of the new model.

ref: 1968

Commit:
[ADD] shopfloor_packing_info: Messages for Checkout Packing Info

This PR adds the option to choose predefined messages for the
parameter "Checkout Packing Info", that is set on the res.partner
and reused in other places. For this, the original text field
shopfloor_packing_info has been transformed into a Many2one that
points to the new model shopfloor.packing.info. Thus, the field
in the res.partner has been renamed to shopfloor_packing_info_id,
that has caused some renames in other files that used it. These
new messages are archived instead of deleted, to keep track of
old messages associated to existing (or archived) res.partners.

This new model, shopfloor.packing.info has been moved to the new
module shopfloor_packing_info. Migration hooks have been provided
to migrate the data, that will happen if the old module is updated
at the same time in which the new module is installed.

The new predefined messages can be created directly in the res.partner
or in the menu Inventory > Settings > Shopfloor.

To migrate the data, the existing data has been migrated by assuming
that the first line is the name while all the content is for the text
of the new model.

guewen commented on shopfloor_packing_info/hooks.py:
```diff
@@ -0,0 +1,131 @@
+        )
+
+
+def _column_exists(cr, table_name, column_name):
```
there is an implementation of this in `odoo.tools.sql.column_exists`

carlosserratoro replied:
I didn't know. Thanks!

guewen commented on shopfloor_packing_info/models/shopfloor_packing_info.py:
```diff
@@ -0,0 +1,20 @@
+    _name = "shopfloor.packing.info"
+    _description = "Shopfloor Packing Information"
+
+    name = fields.Char("Name", required=True)
+    text = fields.Text("Text", required=True)
+    active = fields.Boolean("Active", default=True)
```
```suggestion
    name = fields.Char(required=True)
    text = fields.Text(required=True)
    active = fields.Boolean(default=True)
```

The string is generated from the field name already, better not to set it when it's the same

guewen commented on shopfloor_packing_info/services/checkout.py:
```diff
@@ -0,0 +1,21 @@
+            shopfloor_packing_info = (
+                picking.shopfloor_packing_info_id
+                and picking.shopfloor_packing_info_id.text
+                or ""
```
I recommend to never use the `... and ... or ...` ternary "operator", because it can be very misleading: see the comment on https://docs.python.org/3.3/faq/programming.html#is-there-an-equivalent-of-c-s-ternary-operator (can't be an issue here but a good habit IMO)

guewen commented on shopfloor_packing_info/views/shopfloor_packing_info_views.xml:
```diff
@@ -0,0 +1,81 @@
+            <form>
+                <header>
+                    <field name="active" invisible="1" />
+                    <button
+                        type="object"
+                        name="toggle_active"
+                        string="Archive"
+                        attrs="{'invisible': [('active', '=', False)]}"
+                    />
+                    <button
+                        type="object"
+                        name="toggle_active"
+                        string="Restore"
+                        attrs="{'invisible': [('active', '=', True)]}"
+                    />
```
Since 13.0 the archived button should use this:

```xml
                    <widget
                        name="web_ribbon"
                        title="Archived"
                        bg_color="bg-danger"
                        attrs="{'invisible': [('active', '=', True)]}"
                    />
```

guewen commented on shopfloor_packing_info/views/shopfloor_packing_info_views.xml:
```diff
@@ -0,0 +1,81 @@
+                    string="All"
+                    name="all"
+                    domain="[('active', 'in', [False, True])]"
+                />
```
did you find examples of this filter in core modules?

I find this:

```xml

                <filter
                    string="Archived"
                    name="inactive"
                    domain="[('active','=',False)]"
                />

```

and if I understand how the filter can be useful, but maybe we prefer UX consistency and use the `Archived` filter? (and so when you have this filter you don't need the active field in the tree view as you see either the archived either the unarchived records)

carlosserratoro replied:
I didn't search for any example, I just wanted the same funcionality that are in the scheduled actions by default, but here optional. I can change it by yours.

guewen commented on shopfloor_packing_info/views/shopfloor_packing_info_views.xml:
```diff
@@ -0,0 +1,81 @@
+            (0, 0, {'view_mode': 'tree',
+                    'view_id': ref('shopfloor_packing_info.shopfloor_packing_info_tree_view')}),
+            (0, 0, {'view_mode': 'form',
+                    'view_id': ref('shopfloor_packing_info.shopfloor_packing_info_form_view')})]"
```
you shouldn't need to declare these if you have only one view of each and they'll be the default ones

carlosserratoro replied:
That's the thing: right now there is just one view. By indicating the view we want, we protect against future views. This was done on purpose and I like to do it always this way (sometimes I use an explicit record of ```ir.actions.act_window.view``` for this... )

guewen replied:
Generally you should never need this because the views you reference have a priority of 16, and when we add views that we don't want to be picked by default, we put a greater number such as 99. If you look in official or OCA addon, you'll never see this for default views, unless the views have a priority which would not make them the default.

guewen commented on shopfloor_packing_info/models/shopfloor_packing_info.py:
```diff
@@ -0,0 +1,20 @@
+    def unlink(self):  # pylint: disable=W8106
+        """Archive, not unlink"""
+        self.write({"active": False})
+
+    def toggle_active(self):
+        for record in self:
+            record.active = not record.active
```
```suggestion
```

Already implemented in the base model

guewen reviewed (commented):
Added comments for details

guewen commented on shopfloor_packing_info/models/shopfloor_packing_info.py:
```diff
@@ -0,0 +1,20 @@
+    text = fields.Text("Text", required=True)
+    active = fields.Boolean("Active", default=True)
+
+    def unlink(self):  # pylint: disable=W8106
+        """Archive, not unlink"""
+        self.write({"active": False})
```
I'm not fund of this:

* if the record has been used, you'll have an error message as there is `ondelete='restrict'`, then you can archive it
* if the record has not been used, then I don't see the problem of deleting it (say you just created one by error)
* it goes against the conventions in other models, so is misleading for users

carlosserratoro replied:
I agree. This was not a good idea...

simahawk commented on shopfloor_packing_info/tests/test_shopfloor_packing_info.py:
```diff
@@ -0,0 +1,29 @@
+        )
+
+    def test_unlink(self):
+        partner = self.env["res.partner"].create(
```
use an existing demo partner instead

simahawk commented on shopfloor_packing_info/hooks.py:
```diff
@@ -0,0 +1,131 @@
+        )
+        cr.execute(
+            """
+            DELETE FROM ir_ui_view WHERE id IN (
```
is this needed? The view will be already deleted by the upgrade of shopfloor, no?

carlosserratoro replied:
Shouldn't be needed. I'm going to remove it from here and place it in our own module, because in Travis I'm getting that error that says that a view still has a field that was removed (because of the copy of the view that is stored in the database). I thought that placing it here the problem would be fixed...

simahawk commented on shopfloor_packing_info/hooks.py:
```diff
@@ -0,0 +1,131 @@
+            partner = res_partner.browse(partner_id)
+
+            # We assume the name is the first line, if there are several.
+            linesep_pos = packing_info_full.find(os.linesep)
```
just `packing_info_full.splitlines()`?

carlosserratoro replied:
In this case it doesn't matter because it's a code that is executed once, but I didn't use splitlines nor split because it's much slower. See:

```python
>>> import timeit
>>>
>>> timeit.timeit("x.find('\\n')", setup="x = 'hi\\nhow\\nare\\nyou\\ndoing'")
0.10260158000164665
>>> timeit.timeit("x.find('\\n')", setup="x = 'hi\\nhow\\nare\\nyou\\ndoing'")
0.09811641899796086
>>> timeit.timeit("x.find('\\n')", setup="x = 'hi\\nhow\\nare\\nyou\\ndoing'")
0.09946276299888268
>>> 
>>> timeit.timeit('x.splitlines()', setup="x = 'hi\\nhow\\nare\\nyou\\ndoing'")
0.23253642599956947
>>> timeit.timeit('x.splitlines()', setup="x = 'hi\\nhow\\nare\\nyou\\ndoing'")
0.23402035599792725
>>> timeit.timeit('x.splitlines()', setup="x = 'hi\\nhow\\nare\\nyou\\ndoing'")
0.236937740999565
```

simahawk commented on shopfloor_packing_info/hooks.py:
```diff
@@ -0,0 +1,131 @@
+                ir_model_data._update_xmlids(
+                    [
+                        {
+                            "xml_id": "shopfloor.packing_info_{}".format(xml_id),
```
this xid is wrong. Also, I wouldn't add any xid because you did not have any record before, just strings inserted manually,

carlosserratoro replied:
You are right, the prefix should be ```shopfloor_packing_info```. The XML-id is placed here to ease exporting the existing records to clean them and then importing them in several databases without having to depend on the XML-id assigned by each database. Right now, this is actually not needed, and is thinking about the future when we do this field translatable. If you still don't want it, I can remove it for now...

simahawk replied:
IMO it's useless. If you keep it I'd replace the module name w/ `__export__`

carlosserratoro replied:
Doesn't matter, I'll remove it.

simahawk reviewed (changes requested):
I share all the remarks by Guewen and add few more :wink:

Commit:
[IMP] shopfloor_packing_info: Code review

Changes due to the code review:
- Uses the web_ribbon widget to indicate that a record was archived.
- Unlinking behaviour is back to standard: unlinks. Or, complains if
  the record is used somewhere else.
- X and Y or Z is changed into the safer (and faster) Y if X else Z
- Other minor changes.

carlosserratoro commented:
@simahawk @guewen https://github.com/OCA/wms/pull/161/commits/b3adbf214e16be433c82e03adbb8ed34d3679d2b has been added attending your comments.

guewen commented on shopfloor_packing_info/security/ir.model.access.csv:
```diff
@@ -0,0 +1,4 @@
+"id","name","model_id/id","group_id/id","perm_read","perm_write","perm_create","perm_unlink"
+"access_shopfloor_packing_info_user","shopfloor packing info user","model_shopfloor_packing_info","stock.group_stock_user",1,0,0,0
+"access_shopfloor_packing_info_manager","shopfloor packing info manager","model_shopfloor_packing_info","stock.group_stock_manager",1,1,1,0
```
```suggestion
"access_shopfloor_packing_info_manager","shopfloor packing info manager","model_shopfloor_packing_info","stock.group_stock_manager",1,1,1,1
```

Commit:
[IMP] shopfloor_packing_info: Allow to delete a packing info

Permission has been granted to stock managers to delete
records of shopfloor.packing.info.

sebalix commented on shopfloor_packing_info/demo/shopfloor_packing_info_demo.xml:
```diff
@@ -0,0 +1,11 @@
+        <field name="name">Pallet 1.7m</field>
+        <field name="text">Pallet height limit 1.7m</field>
+    </record>
+</odoo>
```
Given these examples, is it not desirable to have several messages link to a partner, and concat them when needed? Like
```
Max weight per pallet 1000kg
Pallet height limit 1.7m
```
NOTE: but I don't know how this text field was used
